File: src/python/common/core/queue_error_handler.py
"""
Queue Error Handling and Retry System

Provides comprehensive error classification, retry strategies with exponential
backoff, dead letter queue management, and circuit breaker pattern for queue operations.

Features:
    - Error categorization (transient vs permanent)
    - Exponential backoff retry strategy
    - Per-error-type retry limits
    - Dead letter queue for permanently failed items
    - Circuit breaker pattern for repeated failures
    - Error metrics collection and monitoring

Example:
    ```python
    from workspace_qdrant_mcp.core.queue_error_handler import (
        ErrorHandler, ErrorCategory, ErrorType
    )

    handler = ErrorHandler(queue_client)

    # Handle error with automatic categorization
    should_retry = await handler.handle_error(
        file_path="/path/to/file.txt",
        error=exception,
        context={"operation": "ingest"}
    )
    ```
"""

import time
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any

from loguru import logger


class ErrorCategory(Enum):
    """Error category classification."""
    TRANSIENT = "transient"  # Temporary errors, retry possible
    PERMANENT = "permanent"  # Permanent errors, no retry
    RATE_LIMIT = "rate_limit"  # Rate limiting, retry with backoff
    RESOURCE = "resource"  # Resource exhaustion, retry with delay


class ErrorType(Enum):
    """Specific error types with categorization."""

    # Transient errors (retry with backoff)
    NETWORK_TIMEOUT = ("network_timeout", ErrorCategory.TRANSIENT, 5)
    CONNECTION_REFUSED = ("connection_refused", ErrorCategory.TRANSIENT, 5)
    TEMPORARY_FAILURE = ("temporary_failure", ErrorCategory.TRANSIENT, 3)
    DATABASE_LOCKED = ("database_locked", ErrorCategory.TRANSIENT, 10)

    # Rate limit errors (retry with longer backoff)
    RATE_LIMIT_EXCEEDED = ("rate_limit_exceeded", ErrorCategory.RATE_LIMIT, 10)
    TOO_MANY_REQUESTS = ("too_many_requests", ErrorCategory.RATE_LIMIT, 8)

    # Resource errors (retry with delay)
    OUT_OF_MEMORY = ("out_of_memory", ErrorCategory.RESOURCE, 3)
    DISK_FULL = ("disk_full", ErrorCategory.RESOURCE, 3)
    QUOTA_EXCEEDED = ("quota_exceeded", ErrorCategory.RESOURCE, 5)

    # Permanent errors (no retry)
    FILE_NOT_FOUND = ("file_not_found", ErrorCategory.PERMANENT, 0)
    INVALID_FORMAT = ("invalid_format", ErrorCategory.PERMANENT, 0)
    PERMISSION_DENIED = ("permission_denied", ErrorCategory.PERMANENT, 0)
    INVALID_CONFIGURATION = ("invalid_configuration", ErrorCategory.PERMANENT, 0)
    VALIDATION_ERROR = ("validation_error", ErrorCategory.PERMANENT, 0)
    MALFORMED_DATA = ("malformed_data", ErrorCategory.PERMANENT, 0)

    def __init__(self, error_name: str, category: ErrorCategory, max_retries: int):
        self.error_name = error_name
        self.category = category
        self.max_retries = max_retries


@dataclass
class RetryConfig:
    """Retry strategy configuration."""

    # Base retry delay in seconds
    base_delay: float = 1.0

    # Maximum delay between retries
    max_delay: float = 300.0  # 5 minutes

    # Exponential backoff multiplier
    backoff_multiplier: float = 2.0

    # Jitter factor (0-1) to randomize delays
    jitter_factor: float = 0.1

    # Default max retries (when error type not specified)
    default_max_retries: int = 3


@dataclass
class CircuitBreakerConfig:
    """Circuit breaker pattern configuration."""

    # Number of failures before opening circuit
    failure_threshold: int = 5

    # Time window for counting failures (seconds)
    failure_window: int = 60

    # Time to wait before trying half-open state (seconds)
    recovery_timeout: int = 300

    # Success count needed in half-open state to close circuit
    success_threshold: int = 2


@dataclass
class CircuitBreakerState:
    """Circuit breaker state tracking."""

    state: str = "closed"  # closed, open, half-open
    failure_count: int = 0
    success_count: int = 0
    last_failure_time: float | None = None
    opened_at: float | None = None
    failures: list = field(default_factory=list)


@dataclass
class ErrorMetrics:
    """Error metrics for monitoring."""

    total_errors: int = 0
    transient_errors: int = 0
    permanent_errors: int = 0
    rate_limit_errors: int = 0
    resource_errors: int = 0
    retry_count: int = 0
    circuit_breaker_opens: int = 0
    dead_letter_items: int = 0
    successful_retries: int = 0
    failed_retries: int = 0


class ErrorHandler:
    """
    Comprehensive error handler for queue operations.

    Handles error classification, retry strategies, circuit breaker pattern,
    and dead letter queue management.
    """

    def __init__(
        self,
        queue_client,
        retry_config: RetryConfig | None = None,
        circuit_breaker_config: CircuitBreakerConfig | None = None
    ):
        """
        Initialize error handler.

        Args:
            queue_client: SQLiteQueueClient instance
            retry_config: Retry strategy configuration
            circuit_breaker_config: Circuit breaker configuration
        """
        self.queue_client = queue_client
        self.retry_config = retry_config or RetryConfig()
        self.circuit_breaker_config = circuit_breaker_config or CircuitBreakerConfig()

        # Circuit breaker state per collection
        self.circuit_breakers: dict[str, CircuitBreakerState] = {}

        # Error metrics
        self.metrics = ErrorMetrics()

        # Error type mapping from exception messages
        self.error_patterns = self._build_error_patterns()

    def _build_error_patterns(self) -> dict[str, ErrorType]:
        """Build mapping from error message patterns to error types."""
        return {
            "timeout": ErrorType.NETWORK_TIMEOUT,
            "timed out": ErrorType.NETWORK_TIMEOUT,
            "connection refused": ErrorType.CONNECTION_REFUSED,
            "database is locked": ErrorType.DATABASE_LOCKED,
            "rate limit": ErrorType.RATE_LIMIT_EXCEEDED,
            "too many requests": ErrorType.TOO_MANY_REQUESTS,
            "429": ErrorType.TOO_MANY_REQUESTS,
            "out of memory": ErrorType.OUT_OF_MEMORY,
            "disk full": ErrorType.DISK_FULL,
            "no space left": ErrorType.DISK_FULL,
            "quota exceeded": ErrorType.QUOTA_EXCEEDED,
            "file not found": ErrorType.FILE_NOT_FOUND,
            "no such file": ErrorType.FILE_NOT_FOUND,
            "invalid format": ErrorType.INVALID_FORMAT,
            "permission denied": ErrorType.PERMISSION_DENIED,
            "access denied": ErrorType.PERMISSION_DENIED,
            "validation error": ErrorType.VALIDATION_ERROR,
            "malformed": ErrorType.MALFORMED_DATA,
        }

    def classify_error(
        self,
        error: Exception,
        context: dict[str, Any] | None = None
    ) -> ErrorType:
        """
        Classify error by analyzing exception message and context.

        Args:
            error: Exception to classify
            context: Additional context for classification

        Returns:
            ErrorType classification
        """
        error_message = str(error).lower()

        # Check error patterns
        for pattern, error_type in self.error_patterns.items():
            if pattern in error_message:
                logger.debug(f"Classified error as {error_type.error_name}: {pattern}")
                return error_type

        # Default to temporary failure for unknown errors
        logger.warning(f"Unknown error type, classifying as temporary: {error}")
        return ErrorType.TEMPORARY_FAILURE

    def calculate_retry_delay(
        self,
        retry_count: int,
        error_type: ErrorType
    ) -> float:
        """
        Calculate retry delay with exponential backoff and jitter.

        Args:
            retry_count: Current retry attempt number (0-indexed)
            error_type: Type of error

        Returns:
            Delay in seconds before next retry
        """
        # Base delay with exponential backoff
        delay = self.retry_config.base_delay * (
            self.retry_config.backoff_multiplier ** retry_count
        )

        # Apply different multipliers based on error category
        if error_type.category == ErrorCategory.RATE_LIMIT:
            delay *= 3.0  # Longer delays for rate limits
        elif error_type.category == ErrorCategory.RESOURCE:
            delay *= 2.0  # Moderate delays for resource issues

        # Cap at max delay
        delay = min(delay, self.retry_config.max_delay)

        # Add jitter to prevent thundering herd
        import random
        jitter = delay * self.retry_config.jitter_factor * random.random()
        delay += jitter

        logger.debug(
            f"Calculated retry delay for {error_type.error_name}: "
            f"{delay:.2f}s (attempt {retry_count + 1})"
        )

        return delay

    def should_retry(
        self,
        error_type: ErrorType,
        retry_count: int
    ) -> bool:
        """
        Determine if error should be retried.

        Args:
            error_type: Type of error
            retry_count: Current retry count

        Returns:
            True if should retry, False otherwise
        """
        # Permanent errors never retry
        if error_type.category == ErrorCategory.PERMANENT:
            return False

        # Check if max retries exceeded
        max_retries = error_type.max_retries or self.retry_config.default_max_retries

        return retry_count < max_retries

    def get_circuit_breaker(self, collection: str) -> CircuitBreakerState:
        """Get or create circuit breaker state for collection."""
        if collection not in self.circuit_breakers:
            self.circuit_breakers[collection] = CircuitBreakerState()
        return self.circuit_breakers[collection]

    def check_circuit_breaker(self, collection: str) -> tuple[bool, str]:
        """
        Check circuit breaker state.

        Args:
            collection: Collection name

        Returns:
            Tuple of (can_proceed, reason)
        """
        breaker = self.get_circuit_breaker(collection)
        current_time = time.time()

        # If circuit is open, check if recovery timeout has elapsed
        if breaker.state == "open":
            if breaker.opened_at and (
                current_time - breaker.opened_at
            ) > self.circuit_breaker_config.recovery_timeout:
                # Try half-open state
                breaker.state = "half-open"
                breaker.success_count = 0
                logger.info(f"Circuit breaker {collection} entering half-open state")
                return True, "half-open"
            else:
                return False, "circuit_open"

        # Circuit closed or half-open, can proceed
        return True, breaker.state

    def record_circuit_breaker_failure(self, collection: str):
        """Record failure in circuit breaker."""
        breaker = self.get_circuit_breaker(collection)
        current_time = time.time()

        breaker.failure_count += 1
        breaker.last_failure_time = current_time
        breaker.failures.append(current_time)

        # Remove old failures outside time window
        window_start = current_time - self.circuit_breaker_config.failure_window
        breaker.failures = [f for f in breaker.failures if f > window_start]

        # Check if should open circuit
        if breaker.state == "closed":
            if len(breaker.failures) >= self.circuit_breaker_config.failure_threshold:
                breaker.state = "open"
                breaker.opened_at = current_time
                self.metrics.circuit_breaker_opens += 1
                logger.warning(
                    f"Circuit breaker opened for {collection} after "
                    f"{len(breaker.failures)} failures"
                )

        elif breaker.state == "half-open":
            # Failed in half-open, back to open
            breaker.state = "open"
            breaker.opened_at = current_time
            breaker.success_count = 0
            logger.warning(f"Circuit breaker {collection} reopened after half-open failure")

    def record_circuit_breaker_success(self, collection: str):
        """Record success in circuit breaker."""
        breaker = self.get_circuit_breaker(collection)

        if breaker.state == "half-open":
            breaker.success_count += 1

            if breaker.success_count >= self.circuit_breaker_config.success_threshold:
                breaker.state = "closed"
                breaker.failure_count = 0
                breaker.failures = []
                logger.info(f"Circuit breaker {collection} closed after successful recovery")

    async def handle_error(
        self,
        file_path: str,
        error: Exception,
        collection: str,
        context: dict[str, Any] | None = None
    ) -> tuple[bool, ErrorType]:
        """
        Handle error with classification, retry logic, and circuit breaker.

        Args:
            file_path: File path that caused error
            error: Exception that occurred
            collection: Collection name
            context: Additional error context

        Returns:
            Tuple of (should_retry, error_type)
        """
        # Classify error
        error_type = self.classify_error(error, context)

        # Update metrics
        self.metrics.total_errors += 1
        if error_type.category == ErrorCategory.TRANSIENT:
            self.metrics.transient_errors += 1
        elif error_type.category == ErrorCategory.PERMANENT:
            self.metrics.permanent_errors += 1
        elif error_type.category == ErrorCategory.RATE_LIMIT:
            self.metrics.rate_limit_errors += 1
        elif error_type.category == ErrorCategory.RESOURCE:
            self.metrics.resource_errors += 1

        # Check circuit breaker
        can_proceed, breaker_reason = self.check_circuit_breaker(collection)

        if not can_proceed:
            logger.warning(
                f"Circuit breaker open for {collection}, moving to dead letter queue"
            )
            await self.move_to_dead_letter_queue(
                file_path, error_type, str(error), {"circuit_breaker": "open"}
            )
            return False, error_type

        # Get current retry count
        # This would come from the queue item
        retry_count = 0  # Placeholder, should be fetched from queue

        # Check if should retry
        if not self.should_retry(error_type, retry_count):
            logger.info(
                f"Error {error_type.error_name} for {file_path} is not retryable, "
                f"moving to dead letter queue"
            )
            await self.move_to_dead_letter_queue(
                file_path, error_type, str(error), context or {}
            )
            self.record_circuit_breaker_failure(collection)
            return False, error_type

        # Record error and determine retry delay
        max_retries = error_type.max_retries or self.retry_config.default_max_retries

        should_retry_result, error_message_id = await self.queue_client.mark_error(
            file_path=file_path,
            error_type=error_type.error_name,
            error_message=str(error),
            error_details=context,
            max_retries=max_retries
        )

        if should_retry_result:
            self.metrics.retry_count += 1

            # Calculate and apply backoff delay
            delay = self.calculate_retry_delay(retry_count, error_type)
            logger.info(
                f"Retrying {file_path} after {delay:.2f}s "
                f"(attempt {retry_count + 1}/{max_retries})"
            )

            return True, error_type
        else:
            # Max retries exceeded, move to dead letter queue
            logger.warning(
                f"Max retries exceeded for {file_path}, moving to dead letter queue"
            )
            await self.move_to_dead_letter_queue(
                file_path, error_type, str(error), context or {}
            )
            self.record_circuit_breaker_failure(collection)
            return False, error_type

    async def move_to_dead_letter_queue(
        self,
        file_path: str,
        error_type: ErrorType,
        error_message: str,
        context: dict[str, Any]
    ):
        """
        Move permanently failed item to dead letter queue.

        Args:
            file_path: File path
            error_type: Error type
            error_message: Error message
            context: Error context
        """
        # In our design, dead letter queue is represented by removing from
        # ingestion_queue and keeping only the error message
        # For a full implementation, you might want a separate dead_letter_queue table

        error_details = {
            **context,
            "dead_letter_reason": error_type.error_name,
            "moved_to_dlq_at": datetime.utcnow().isoformat(),
        }

        # Record in messages table
        await self.queue_client.mark_error(
            file_path=file_path,
            error_type=f"DEAD_LETTER_{error_type.error_name}",
            error_message=error_message,
            error_details=error_details,
            max_retries=0  # Force removal from queue
        )

        self.metrics.dead_letter_items += 1

        logger.error(
            f"Moved {file_path} to dead letter queue: {error_type.error_name}"
        )

    def get_metrics(self) -> dict[str, Any]:
        """
        Get error handling metrics.

        Returns:
            Dictionary of metrics
        """
        return {
            "total_errors": self.metrics.total_errors,
            "transient_errors": self.metrics.transient_errors,
            "permanent_errors": self.metrics.permanent_errors,
            "rate_limit_errors": self.metrics.rate_limit_errors,
            "resource_errors": self.metrics.resource_errors,
            "retry_count": self.metrics.retry_count,
            "circuit_breaker_opens": self.metrics.circuit_breaker_opens,
            "dead_letter_items": self.metrics.dead_letter_items,
            "successful_retries": self.metrics.successful_retries,
            "failed_retries": self.metrics.failed_retries,
            "active_circuit_breakers": {
                name: state.state
                for name, state in self.circuit_breakers.items()
                if state.state != "closed"
            }
        }

    def reset_metrics(self):
        """Reset error metrics."""
        self.metrics = ErrorMetrics()
        logger.info("Error metrics reset")

File: src/rust/daemon/core/src/queue_error_handler.rs
// Queue Error Handling and Retry System
//
// Provides comprehensive error classification, retry strategies with exponential
// backoff, dead letter queue management, and circuit breaker pattern for queue operations.

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};
use rand::Rng;
use serde::{Deserialize, Serialize};
use tracing::{debug, error, info, warn};

use crate::queue_operations::QueueManager;

/// Error category classification
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ErrorCategory {
    /// Temporary errors, retry possible
    Transient,
    /// Permanent errors, no retry
    Permanent,
    /// Rate limiting, retry with backoff
    RateLimit,
    /// Resource exhaustion, retry with delay
    Resource,
}

/// Specific error types with categorization
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ErrorType {
    // Transient errors (retry with backoff)
    NetworkTimeout,
    ConnectionRefused,
    TemporaryFailure,
    DatabaseLocked,

    // Rate limit errors (retry with longer backoff)
    RateLimitExceeded,
    TooManyRequests,

    // Resource errors (retry with delay)
    OutOfMemory,
    DiskFull,
    QuotaExceeded,

    // Permanent errors (no retry)
    FileNotFound,
    InvalidFormat,
    PermissionDenied,
    InvalidConfiguration,
    ValidationError,
    MalformedData,
}

impl ErrorType {
    pub fn category(&self) -> ErrorCategory {
        match self {
            ErrorType::NetworkTimeout
            | ErrorType::ConnectionRefused
            | ErrorType::TemporaryFailure
            | ErrorType::DatabaseLocked => ErrorCategory::Transient,

            ErrorType::RateLimitExceeded | ErrorType::TooManyRequests => {
                ErrorCategory::RateLimit
            }

            ErrorType::OutOfMemory | ErrorType::DiskFull | ErrorType::QuotaExceeded => {
                ErrorCategory::Resource
            }

            ErrorType::FileNotFound
            | ErrorType::InvalidFormat
            | ErrorType::PermissionDenied
            | ErrorType::InvalidConfiguration
            | ErrorType::ValidationError
            | ErrorType::MalformedData => ErrorCategory::Permanent,
        }
    }

    pub fn max_retries(&self) -> i32 {
        match self {
            ErrorType::NetworkTimeout => 5,
            ErrorType::ConnectionRefused => 5,
            ErrorType::TemporaryFailure => 3,
            ErrorType::DatabaseLocked => 10,
            ErrorType::RateLimitExceeded => 10,
            ErrorType::TooManyRequests => 8,
            ErrorType::OutOfMemory => 3,
            ErrorType::DiskFull => 3,
            ErrorType::QuotaExceeded => 5,
            _ => 0, // Permanent errors
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            ErrorType::NetworkTimeout => "network_timeout",
            ErrorType::ConnectionRefused => "connection_refused",
            ErrorType::TemporaryFailure => "temporary_failure",
            ErrorType::DatabaseLocked => "database_locked",
            ErrorType::RateLimitExceeded => "rate_limit_exceeded",
            ErrorType::TooManyRequests => "too_many_requests",
            ErrorType::OutOfMemory => "out_of_memory",
            ErrorType::DiskFull => "disk_full",
            ErrorType::QuotaExceeded => "quota_exceeded",
            ErrorType::FileNotFound => "file_not_found",
            ErrorType::InvalidFormat => "invalid_format",
            ErrorType::PermissionDenied => "permission_denied",
            ErrorType::InvalidConfiguration => "invalid_configuration",
            ErrorType::ValidationError => "validation_error",
            ErrorType::MalformedData => "malformed_data",
        }
    }
}

/// Retry strategy configuration
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Base retry delay in seconds
    pub base_delay: f64,
    /// Maximum delay between retries
    pub max_delay: f64,
    /// Exponential backoff multiplier
    pub backoff_multiplier: f64,
    /// Jitter factor (0-1) to randomize delays
    pub jitter_factor: f64,
    /// Default max retries (when error type not specified)
    pub default_max_retries: i32,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            base_delay: 1.0,
            max_delay: 300.0,  // 5 minutes
            backoff_multiplier: 2.0,
            jitter_factor: 0.1,
            default_max_retries: 3,
        }
    }
}

/// Circuit breaker pattern configuration
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    /// Number of failures before opening circuit
    pub failure_threshold: usize,
    /// Time window for counting failures (seconds)
    pub failure_window: u64,
    /// Time to wait before trying half-open state (seconds)
    pub recovery_timeout: u64,
    /// Success count needed in half-open state to close circuit
    pub success_threshold: usize,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: 5,
            failure_window: 60,
            recovery_timeout: 300,
            success_threshold: 2,
        }
    }
}

/// Circuit breaker state
#[derive(Debug, Clone)]
pub enum CircuitState {
    Closed,
    Open,
    HalfOpen,
}

impl CircuitState {
    pub fn as_str(&self) -> &'static str {
        match self {
            CircuitState::Closed => "closed",
            CircuitState::Open => "open",
            CircuitState::HalfOpen => "half-open",
        }
    }
}

/// Circuit breaker state tracking
#[derive(Debug, Clone)]
pub struct CircuitBreakerState {
    pub state: CircuitState,
    pub failure_count: usize,
    pub success_count: usize,
    pub last_failure_time: Option<u64>,
    pub opened_at: Option<u64>,
    pub failures: Vec<u64>,
}

impl Default for CircuitBreakerState {
    fn default() -> Self {
        Self {
            state: CircuitState::Closed,
            failure_count: 0,
            success_count: 0,
            last_failure_time: None,
            opened_at: None,
            failures: Vec::new(),
        }
    }
}

/// Error metrics for monitoring
#[derive(Debug, Clone, Default, Serialize)]
pub struct ErrorMetrics {
    pub total_errors: u64,
    pub transient_errors: u64,
    pub permanent_errors: u64,
    pub rate_limit_errors: u64,
    pub resource_errors: u64,
    pub retry_count: u64,
    pub circuit_breaker_opens: u64,
    pub dead_letter_items: u64,
    pub successful_retries: u64,
    pub failed_retries: u64,
}

/// Comprehensive error handler for queue operations
pub struct ErrorHandler {
    queue_manager: QueueManager,
    retry_config: RetryConfig,
    circuit_breaker_config: CircuitBreakerConfig,
    circuit_breakers: HashMap<String, CircuitBreakerState>,
    metrics: ErrorMetrics,
}

impl ErrorHandler {
    /// Create new error handler
    pub fn new(
        queue_manager: QueueManager,
        retry_config: Option<RetryConfig>,
        circuit_breaker_config: Option<CircuitBreakerConfig>,
    ) -> Self {
        Self {
            queue_manager,
            retry_config: retry_config.unwrap_or_default(),
            circuit_breaker_config: circuit_breaker_config.unwrap_or_default(),
            circuit_breakers: HashMap::new(),
            metrics: ErrorMetrics::default(),
        }
    }

    /// Classify error by analyzing error message
    pub fn classify_error(&self, error_message: &str) -> ErrorType {
        let error_lower = error_message.to_lowercase();

        if error_lower.contains("timeout") || error_lower.contains("timed out") {
            ErrorType::NetworkTimeout
        } else if error_lower.contains("connection refused") {
            ErrorType::ConnectionRefused
        } else if error_lower.contains("database is locked") {
            ErrorType::DatabaseLocked
        } else if error_lower.contains("rate limit") {
            ErrorType::RateLimitExceeded
        } else if error_lower.contains("too many requests") || error_lower.contains("429") {
            ErrorType::TooManyRequests
        } else if error_lower.contains("out of memory") {
            ErrorType::OutOfMemory
        } else if error_lower.contains("disk full") || error_lower.contains("no space left") {
            ErrorType::DiskFull
        } else if error_lower.contains("quota exceeded") {
            ErrorType::QuotaExceeded
        } else if error_lower.contains("file not found") || error_lower.contains("no such file") {
            ErrorType::FileNotFound
        } else if error_lower.contains("invalid format") {
            ErrorType::InvalidFormat
        } else if error_lower.contains("permission denied")
            || error_lower.contains("access denied")
        {
            ErrorType::PermissionDenied
        } else if error_lower.contains("validation error") {
            ErrorType::ValidationError
        } else if error_lower.contains("malformed") {
            ErrorType::MalformedData
        } else {
            warn!("Unknown error type, classifying as temporary: {}", error_message);
            ErrorType::TemporaryFailure
        }
    }

    /// Calculate retry delay with exponential backoff and jitter
    pub fn calculate_retry_delay(&self, retry_count: i32, error_type: ErrorType) -> f64 {
        // Base delay with exponential backoff
        let mut delay = self.retry_config.base_delay
            * self.retry_config.backoff_multiplier.powi(retry_count);

        // Apply different multipliers based on error category
        match error_type.category() {
            ErrorCategory::RateLimit => delay *= 3.0,  // Longer delays for rate limits
            ErrorCategory::Resource => delay *= 2.0,   // Moderate delays for resource issues
            _ => {}
        }

        // Cap at max delay
        delay = delay.min(self.retry_config.max_delay);

        // Add jitter to prevent thundering herd
        let mut rng = rand::thread_rng();
        let jitter = delay * self.retry_config.jitter_factor * rng.gen::<f64>();
        delay += jitter;

        debug!(
            "Calculated retry delay for {}: {:.2}s (attempt {})",
            error_type.as_str(),
            delay,
            retry_count + 1
        );

        delay
    }

    /// Determine if error should be retried
    pub fn should_retry(&self, error_type: ErrorType, retry_count: i32) -> bool {
        // Permanent errors never retry
        if error_type.category() == ErrorCategory::Permanent {
            return false;
        }

        // Check if max retries exceeded
        let max_retries = if error_type.max_retries() > 0 {
            error_type.max_retries()
        } else {
            self.retry_config.default_max_retries
        };

        retry_count < max_retries
    }

    /// Get or create circuit breaker state for collection
    pub fn get_circuit_breaker(&mut self, collection: &str) -> &mut CircuitBreakerState {
        self.circuit_breakers
            .entry(collection.to_string())
            .or_default()
    }

    /// Check circuit breaker state
    pub fn check_circuit_breaker(&mut self, collection: &str) -> (bool, &str) {
        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        let recovery_timeout = self.circuit_breaker_config.recovery_timeout;

        let breaker = self.get_circuit_breaker(collection);

        match breaker.state {
            CircuitState::Open => {
                if let Some(opened_at) = breaker.opened_at {
                    if (current_time - opened_at) > recovery_timeout {
                        // Try half-open state
                        breaker.state = CircuitState::HalfOpen;
                        breaker.success_count = 0;
                        info!("Circuit breaker {} entering half-open state", collection);
                        return (true, "half-open");
                    }
                }
                (false, "circuit_open")
            }
            _ => (true, breaker.state.as_str()),
        }
    }

    /// Record failure in circuit breaker
    pub fn record_circuit_breaker_failure(&mut self, collection: &str) {
        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let failure_window = self.circuit_breaker_config.failure_window;
        let failure_threshold = self.circuit_breaker_config.failure_threshold;

        let breaker = self.get_circuit_breaker(collection);
        breaker.failure_count += 1;
        breaker.last_failure_time = Some(current_time);
        breaker.failures.push(current_time);

        // Remove old failures outside time window
        let window_start = current_time - failure_window;
        breaker.failures.retain(|&f| f > window_start);

        // Check if should open circuit
        let failure_count = breaker.failures.len();
        match breaker.state {
            CircuitState::Closed => {
                if failure_count >= failure_threshold {
                    breaker.state = CircuitState::Open;
                    breaker.opened_at = Some(current_time);
                    self.metrics.circuit_breaker_opens += 1;
                    warn!(
                        "Circuit breaker opened for {} after {} failures",
                        collection,
                        failure_count
                    );
                }
            }
            CircuitState::HalfOpen => {
                // Failed in half-open, back to open
                breaker.state = CircuitState::Open;
                breaker.opened_at = Some(current_time);
                breaker.success_count = 0;
                warn!("Circuit breaker {} reopened after half-open failure", collection);
            }
            _ => {}
        }
    }

    /// Record success in circuit breaker
    pub fn record_circuit_breaker_success(&mut self, collection: &str) {
        let success_threshold = self.circuit_breaker_config.success_threshold;
        let breaker = self.get_circuit_breaker(collection);

        if matches!(breaker.state, CircuitState::HalfOpen) {
            breaker.success_count += 1;

            if breaker.success_count >= success_threshold {
                breaker.state = CircuitState::Closed;
                breaker.failure_count = 0;
                breaker.failures.clear();
                info!("Circuit breaker {} closed after successful recovery", collection);
            }
        }
    }

    /// Handle error with classification, retry logic, and circuit breaker
    pub async fn handle_error(
        &mut self,
        file_path: &str,
        error_message: &str,
        collection: &str,
        context: Option<&HashMap<String, serde_json::Value>>,
    ) -> (bool, ErrorType) {
        // Classify error
        let error_type = self.classify_error(error_message);

        // Update metrics
        self.metrics.total_errors += 1;
        match error_type.category() {
            ErrorCategory::Transient => self.metrics.transient_errors += 1,
            ErrorCategory::Permanent => self.metrics.permanent_errors += 1,
            ErrorCategory::RateLimit => self.metrics.rate_limit_errors += 1,
            ErrorCategory::Resource => self.metrics.resource_errors += 1,
        }

        // Check circuit breaker
        let (can_proceed, _breaker_reason) = self.check_circuit_breaker(collection);

        if !can_proceed {
            warn!(
                "Circuit breaker open for {}, moving to dead letter queue",
                collection
            );
            self.move_to_dead_letter_queue(
                file_path,
                error_type,
                error_message,
                Some(&HashMap::from([("circuit_breaker".to_string(), serde_json::json!("open"))])),
            )
            .await;
            return (false, error_type);
        }

        // Get current retry count (placeholder, should be fetched from queue)
        let retry_count = 0;

        // Check if should retry
        if !self.should_retry(error_type, retry_count) {
            info!(
                "Error {} for {} is not retryable, moving to dead letter queue",
                error_type.as_str(),
                file_path
            );
            self.move_to_dead_letter_queue(file_path, error_type, error_message, context)
                .await;
            self.record_circuit_breaker_failure(collection);
            return (false, error_type);
        }

        // Record error and determine retry delay
        let max_retries = if error_type.max_retries() > 0 {
            error_type.max_retries()
        } else {
            self.retry_config.default_max_retries
        };

        let result = self
            .queue_manager
            .mark_error(
                file_path,
                error_type.as_str(),
                error_message,
                context,
                max_retries,
            )
            .await;

        match result {
            Ok((should_retry_result, _error_message_id)) => {
                if should_retry_result {
                    self.metrics.retry_count += 1;

                    let delay = self.calculate_retry_delay(retry_count, error_type);
                    info!(
                        "Retrying {} after {:.2}s (attempt {}/{})",
                        file_path,
                        delay,
                        retry_count + 1,
                        max_retries
                    );

                    (true, error_type)
                } else {
                    warn!(
                        "Max retries exceeded for {}, moving to dead letter queue",
                        file_path
                    );
                    self.move_to_dead_letter_queue(file_path, error_type, error_message, context)
                        .await;
                    self.record_circuit_breaker_failure(collection);
                    (false, error_type)
                }
            }
            Err(e) => {
                error!("Failed to record error: {}", e);
                (false, error_type)
            }
        }
    }

    /// Move permanently failed item to dead letter queue
    async fn move_to_dead_letter_queue(
        &mut self,
        file_path: &str,
        error_type: ErrorType,
        error_message: &str,
        context: Option<&HashMap<String, serde_json::Value>>,
    ) {
        let mut error_details = context.cloned().unwrap_or_default();
        error_details.insert(
            "dead_letter_reason".to_string(),
            serde_json::json!(error_type.as_str()),
        );
        error_details.insert(
            "moved_to_dlq_at".to_string(),
            serde_json::json!(wqm_common::timestamps::now_utc()),
        );

        // Record in messages table
        let _ = self
            .queue_manager
            .mark_error(
                file_path,
                &format!("DEAD_LETTER_{}", error_type.as_str()),
                error_message,
                Some(&error_details),
                0, // Force removal from queue
            )
            .await;

        self.metrics.dead_letter_items += 1;

        error!(
            "Moved {} to dead letter queue: {}",
            file_path,
            error_type.as_str()
        );
    }

    /// Get error handling metrics
    pub fn get_metrics(&self) -> HashMap<String, serde_json::Value> {
        let mut metrics = HashMap::new();
        metrics.insert("total_errors".to_string(), serde_json::json!(self.metrics.total_errors));
        metrics.insert("transient_errors".to_string(), serde_json::json!(self.metrics.transient_errors));
        metrics.insert("permanent_errors".to_string(), serde_json::json!(self.metrics.permanent_errors));
        metrics.insert("rate_limit_errors".to_string(), serde_json::json!(self.metrics.rate_limit_errors));
        metrics.insert("resource_errors".to_string(), serde_json::json!(self.metrics.resource_errors));
        metrics.insert("retry_count".to_string(), serde_json::json!(self.metrics.retry_count));
        metrics.insert("circuit_breaker_opens".to_string(), serde_json::json!(self.metrics.circuit_breaker_opens));
        metrics.insert("dead_letter_items".to_string(), serde_json::json!(self.metrics.dead_letter_items));
        metrics.insert("successful_retries".to_string(), serde_json::json!(self.metrics.successful_retries));
        metrics.insert("failed_retries".to_string(), serde_json::json!(self.metrics.failed_retries));

        let active_breakers: HashMap<String, &str> = self
            .circuit_breakers
            .iter()
            .filter(|(_, state)| !matches!(state.state, CircuitState::Closed))
            .map(|(name, state)| (name.clone(), state.state.as_str()))
            .collect();
        metrics.insert("active_circuit_breakers".to_string(), serde_json::json!(active_breakers));

        metrics
    }

    /// Reset error metrics
    pub fn reset_metrics(&mut self) {
        self.metrics = ErrorMetrics::default();
        info!("Error metrics reset");
    }
}
